{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing needed packages \n",
    "import csv \n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paths to csv files storing DeSeq2 data \n",
    "# this may need to be changed\n",
    "pathToGuideData = '/home/data/refined/bc_dcis/EJM_Data_Output/Dixit_Analysis/Dixit_N4_Pseudobulk_Count_DeSeq2_Analysis.Controls_vs_SingleGuide.csv'\n",
    "#pathToGeneData = '/home/data/refined/bc_dcis/EJM_Data_Output/Dixit_Analysis/Dixit_N4_Pseudobulk_Count_DeSeq2_Analysis.JustControls_vs_SingleTargets.csv'\n",
    "pathToGeneData = '/home/data/refined/bc_dcis/EJM_Data_Output/Dixit_Analysis/Dixit_N4_Pseudobulk_Count_DeSeq2_Analysis.Controls_vs_SingleTargets.csv' \n",
    "pathToGeneExpandedData = '/home/data/refined/bc_dcis/EJM_Data_Output/Dixit_Analysis/Dixit_N4_Pseudobulk_Count_DeSeq2_Analysis.Controls_vs_SingleTargets.csv' \n",
    "pathToDualGeneData = '/home/data/refined/bc_dcis/EJM_Data_Output/Dixit_Analysis/Dixit_N4_Pseudobulk_Count_DeSeq2_Analysis.DualIntergenic_vs_MultiTargets.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getter method to get a list of all target genes\n",
    "# for the deseq2 by gene comparisions, there is a target gene column\n",
    "# in the case of the dixit data, this was one of 10 genes encoding a transcription factor \n",
    "# this method returns a list of all unqiue target genes\n",
    "\n",
    "def genes():\n",
    "    \n",
    "    unique = []\n",
    "    fileToRead = pathToGeneData \n",
    "\n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "\n",
    "            if (row[6] != 'NA') and (row[6] != 'perturbation'): # index 6 because the target genes are stored in the 7th column of the table, ignore first header row\n",
    "                if row[6] not in unique:\n",
    "                    unique.append(row[6])\n",
    "    \n",
    "    return(unique)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getter method to list all unique dual-perturbatios\n",
    "# in the dual guide cells only comparisons the 7th column contains different combinations of target genes\n",
    "# this method finds all unqiue dual-gene-target perturbations and returns a list storing them \n",
    "\n",
    "def dualGenes():\n",
    "    \n",
    "    dualGenes = []\n",
    "    fileToRead = pathToDualGeneData\n",
    "\n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "\n",
    "            if (row[6] != 'NA') and (row[6] != 'perturbation'):\n",
    "                if row[6] not in dualGenes:\n",
    "                    dualGenes.append(row[6])\n",
    "    \n",
    "    return(dualGenes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function allows you to specify a gene and see if there are any dual perturbation defined for it\n",
    "# i.e. what are the dual perturbations that include a perturbation to parameter gene \n",
    "# this function returns a list of all found dual-gene-target perturbations\n",
    "\n",
    "def findDual(gene):\n",
    "\n",
    "    found = []\n",
    "    dualList = dualGenes()\n",
    "    for element in dualList:\n",
    "        if gene in element:\n",
    "            found.append(element)\n",
    "\n",
    "    return(found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given a gene* name this function will return lists of which genes get upregulated and downregulated when gene* is perturbed \n",
    "# this is only considering the target gene not specific perturbation \n",
    " \n",
    "def geneTarget(targetGene):\n",
    "\n",
    "    upReg = []\n",
    "    downReg = [] \n",
    "\n",
    "    fileToRead = pathToGeneData\n",
    "\n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "\n",
    "            if row[6] == targetGene: #row of index 6 stores target genes, if it matches the passed parameter, we add the log2_fc (stores in row index 3) to one of two lists\n",
    "                if float(row[3]) < 0: #negative log2_fc values means the perturbation (row index 1) and log2_fc value should be appended to the down regulation list\n",
    "                    downReg.append((row[1], row[3]))\n",
    "                if float(row[3]) > 0: #postitive log2_fc values means the perturbation (row index 1) and log2_fc value should be appended to the up regulation list\n",
    "                    upReg.append((row[1], row[3]))\n",
    "\n",
    "    if (len(upReg) == 0) and (len(downReg) == 0):\n",
    "        return 0 # if the passed parameter does not correspond to a defind target gene the function returns 0\n",
    "    \n",
    "    else: \n",
    "        return ((upReg), (downReg)) # returns lists as a tuple so we can store this information after a function call in a variable \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a method to get a list of all perturbations specified in the data file for a specified target gene\n",
    "# the gene parameter means 'list all perturbations that target [gene]' \n",
    "\n",
    "def perturbations(gene):\n",
    "    \n",
    "    listOfPerturbations = []\n",
    "    fileToRead = pathToGuideData\n",
    "    \n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "\n",
    "            if gene in row[6]: #\n",
    "                if row[6] not in listOfPerturbations:\n",
    "                    listOfPerturbations.append(row[6])\n",
    "    \n",
    "    \n",
    "    return(listOfPerturbations) # returns a list storing the perturbations that target the gene passed as the parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given a guide* name this function will return a list of which genes get upregulated and downregulated when the perturbation is caused by guide* \n",
    "# i.e., when  we have guide x, returns a list of genes that get upregulated, and a list of genes that get downregulated as a result of having that guide/perturbation present\n",
    "\n",
    "def guideImpact(guideRNA):\n",
    "\n",
    "    upRegP = []\n",
    "    downRegP = [] \n",
    "\n",
    "    fileToRead = pathToGuideData\n",
    "\n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "\n",
    "            if row[6] == guideRNA: # column index 6 stores different perturbations, if the perturbation matches the passed parameter, we add the name of the gene (that was differentially expressed) to a list\n",
    "                if float(row[3]) < 0: # if the log2_fc is negative, it means the item should be added to the down regulated list\n",
    "                    downRegP.append((row[1], row[3]))\n",
    "                if float(row[3]) > 0: # if the log2_fc is positive, it means the item should be added to the up regulated list\n",
    "                    upRegP.append((row[1], row[3]))\n",
    "\n",
    "    if (len(upRegP) == 0) and (len(downRegP) == 0):\n",
    "        return 0\n",
    "    \n",
    "    else: \n",
    "        return ((upRegP), (downRegP)) #returns lists as a tuple so we can store this information after a function call in a variable \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "# same function as above but specify 2 target genes\n",
    "# the parameter is a dual gene perturbation (you can print out all the different dual gene perturbations defined in the file using the dualGenes() method \n",
    "# when we have perturbation dualPerturb, it returns lists of up and down regulated genes\n",
    "\n",
    "def dualTarget(dualPerturb):\n",
    "    \n",
    "    if dualPerturb != dualPerturb:\n",
    "        print(\"The dual-pertubation passed is not found in the data file\")\n",
    "        return 1\n",
    "    \n",
    "    upReg = []\n",
    "    downReg = []\n",
    "\n",
    "    fileToRead = pathToDualGeneData\n",
    "    with open(fileToRead, 'r') as file:\n",
    "        csv_reader = csv.reader(file)\n",
    "        for row in csv_reader:\n",
    "            if row[6] == dualPerturb:\n",
    "                if float(row[3]) < 0:\n",
    "                    downReg.append((row[1], row[3]))\n",
    "                if float(row[3]) > 0:\n",
    "                    upReg.append((row[1], row[3]))\n",
    "    \n",
    "    return((upReg, downReg))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function that compares 2 lists \n",
    "# helper method for compareM method\n",
    "# creates 3 lists: 1 that contains elements only in the first list\n",
    "# 1 that contains elements only in the second list\n",
    "# 1 that contains elements that are in both lists \n",
    "\n",
    "def compare(list1, list2):\n",
    "\n",
    "    unique_list1 = []\n",
    "    unique_list2 = []\n",
    "    same_list = []\n",
    "    \n",
    "    # Check for unique tuples in list1\n",
    "    for tup1 in list1:\n",
    "        unique = True\n",
    "        for tup2 in list2:\n",
    "            if tup1[0] == tup2[0]:\n",
    "                unique = False\n",
    "                break\n",
    "        if unique:\n",
    "            unique_list1.append(tup1)\n",
    "    \n",
    "    # Check for unique tuples in list2\n",
    "    for tup2 in list2:\n",
    "        unique = True\n",
    "        for tup1 in list1:\n",
    "            if tup2[0] == tup1[0]:\n",
    "                unique = False\n",
    "                break\n",
    "        if unique:\n",
    "            unique_list2.append(tup2)\n",
    "        else:\n",
    "            same_list.append((tup1[0], tup1[1], tup2[1]))\n",
    "\n",
    "    return ((unique_list1), (unique_list2), (same_list)) #returns lists of unique genes in the first parameter, list of unique genes in the second parameter, and common genes in both parameters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mutation1 is going to be a specific gene or perturbation\n",
    "# if mutation 1 is a specific perturbation put type1 as 'p', if it is a target gene put 't', and if you are intrested in the upregulated list of this mutation put direction as 'u' (or 'd' for downregulated list)\n",
    "# mutation2 is going to be a specific gene or perturbatipn\n",
    "# if mutation 2 is a speciifc perturbation put type2 as 'p', if it is a target gene put 't', and if you are intrested in the upregulated list of this mutation put direction as 'u' (or 'd' for downregulated list)\n",
    "\n",
    "def compareM(mutation1, type1, direction1, mutation2, type2, direction2):\n",
    "    \n",
    "    mut1U, mut1D, mut2U, mut2D = [], [], [], []\n",
    "    u1, u2, s = [], [], []\n",
    "   \n",
    "    #list storing information about first mutation\n",
    "    if type1 == 'p':\n",
    "        \n",
    "        mut1U = (guideImpact(mutation1))[0]\n",
    "\n",
    "          \n",
    "        \n",
    "        mut1D = (guideImpact(mutation1))[1]\n",
    "       \n",
    "\n",
    "    elif type1 == 't':\n",
    "        mut1U = (geneTarget(mutation1))[0]\n",
    "       \n",
    "\n",
    "        mut1D = (geneTarget(mutation1))[1]\n",
    "        \n",
    "\n",
    "    else: \n",
    "        print(\"error: invalid type entered\")\n",
    "        return 0\n",
    "\n",
    "    #list storing information about second mutation\n",
    "    if type2 == 'p':\n",
    "        mut2U = (guideImpact(mutation2))[0]\n",
    "        \n",
    "\n",
    "        mut2D = (guideImpact(mutation2))[1]\n",
    "        \n",
    "\n",
    "    elif type2 == 't':\n",
    "        mut2U = (geneTarget(mutation2))[0]\n",
    "       \n",
    "\n",
    "        mut2D = (geneTarget(mutation2))[1]\n",
    "        \n",
    "            \n",
    "    else: \n",
    "        print(\"error: invalid type entered\")\n",
    "        return 0\n",
    "\n",
    "\n",
    "    # case1: type1 = u type2 = u\n",
    "    # comparing mutation1 upregulated genes and mutation2 upregulated genes\n",
    "    if (direction1 == 'u') and (direction2 == 'u'):\n",
    "        u1, u2, s = compare(mut1U, mut2U)[0], compare(mut1U, mut2U)[1], compare(mut1U, mut2U)[2]\n",
    "        \n",
    "\n",
    "    # case2: type1 = u type2 = d\n",
    "    # comparing mutation1 upregulated genes and mutation2 downregulated genes \n",
    "    elif (direction1 == 'u') and (direction2 == 'd'):\n",
    "        u1, u2, s = compare(mut1U, mut2D)[0], compare(mut1U, mut2D)[1], compare(mut1U, mut2D)[2]\n",
    "        \n",
    "\n",
    "    # case3: type1 = d type2 = d\n",
    "    # comparing mutation1 downregulated genes and mutation2 downregulated genes\n",
    "    elif (direction1 == 'd') and (direction2 == 'd'):\n",
    "        u1, u2, s = compare(mut1D, mut2D)[0], compare(mut1D, mut2D)[1], compare(mut1D, mut2D)[2]\n",
    "        \n",
    "\n",
    "    # case4: type1 = d type2 = u\n",
    "    # comparing mutation1 downregulated genes and mutation2 upregulated genes \n",
    "    elif (direction1 == 'd') and (direction2 == 'u'):\n",
    "        u1, u2, s = compare(mut1D, mut2U)[0], compare(mut1D, mut2U)[1], compare(mut1D, mut2U)[2]\n",
    "        \n",
    "    return((u1), (u2), (s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# log2_fc convert to ratio helper function\n",
    "def convert_log2fc_to_ratio(lst):\n",
    "    converted_lst = []\n",
    "    for tup in lst:\n",
    "        gene_id, log2fc = tup\n",
    "        ratio = 2 ** float(log2fc)\n",
    "        converted_tup = (gene_id, ratio)\n",
    "        converted_lst.append(converted_tup)\n",
    "    return converted_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ratio convert to log2_fc helper function\n",
    "def convert_ratio_to_log2fc(lst):\n",
    " \n",
    "    converted_lst = []\n",
    "    for tup in lst:\n",
    "        gene_id, ratio = tup\n",
    "        log2fc = math.log2(ratio)\n",
    "        log2fc = round(log2fc, 4)\n",
    "        converted_tup = (gene_id, log2fc)\n",
    "        converted_lst.append(converted_tup)\n",
    "    return converted_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "#helper method\n",
    "def listMerger(list1, list2):\n",
    "\n",
    "    merged_dict = {}\n",
    "\n",
    "    # Merge list1 into the dictionary\n",
    "    for item in list1:\n",
    "        key, value = item\n",
    "        if key in merged_dict:\n",
    "            merged_dict[key] *= float(value)\n",
    "        else:\n",
    "            merged_dict[key] = float(value)\n",
    "\n",
    "    # Merge list2 into the dictionary\n",
    "    for item in list2:\n",
    "        key, value = item\n",
    "        if key in merged_dict:\n",
    "            merged_dict[key] *= float(value)\n",
    "        else:\n",
    "            merged_dict[key] = float(value)\n",
    "\n",
    "    # Convert the dictionary back into a list of tuples\n",
    "    merged_list = [(key, value) for key, value in merged_dict.items()]\n",
    "\n",
    "    return merged_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "#finding genes that get upregulated in one situation, but downregulated in another \n",
    "def match(list1, list2):\n",
    "   \n",
    "    match = []\n",
    "    for element in list1:\n",
    "        gene = element[0]\n",
    "        for element2 in list2:\n",
    "            if gene == element2[0]:\n",
    "                found = (gene, element[1], element2[1])\n",
    "                match.append(found)\n",
    "    \n",
    "    return match\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper method to create a theoretical merged list \n",
    "# it will access the up and down regulated lists for each of perturbation in the dual-perturbation pair\n",
    "# it will merge the up and down regulated lists so we end up with one upregulated list and one down regulated list\n",
    "# any shared genes in each list will have their log2_fc added together\n",
    "# the two lists will be compared to ensure no gene appears in both the up and down regulated list, if it does (i.e. shows up in the shared list) the log2_fc get added together\n",
    "# move any genes with a postive log2_fc to the upregulated list, move any genes with a negative log2_fc to the downregulated list\n",
    "# add the log values, this will show the theoretical gene expression we should see when these two perturbations are comnbinded \n",
    "\n",
    "def merge(target1, target2):\n",
    "\n",
    "    target1UP, target1DOWN = [], []\n",
    "    target2UP, target2DOWN = [], []\n",
    "\n",
    "    # we are accessing the 2 up regulated and 2 down regulated lists of each of the targets\n",
    "    # the log2_fc must get converted back to a ratio so if there are any duplicated values we can add them \n",
    "    target1UP = geneTarget(target1)[0]\n",
    "    target1UP = convert_log2fc_to_ratio(target1UP)\n",
    "    target1DOWN = geneTarget(target1)[1]\n",
    "    target1DOWN = convert_log2fc_to_ratio(target1DOWN)\n",
    "    target2UP = geneTarget(target2)[0]\n",
    "    target2UP = convert_log2fc_to_ratio(target2UP)\n",
    "    target2DOWN = geneTarget(target2)[1]\n",
    "    target2DOWN = convert_log2fc_to_ratio(target2DOWN)\n",
    "\n",
    "    # create a theoretical up regulated and down regulate list of genes if effects of the perturbations were purely linear\n",
    "    mergedUp = listMerger(target1UP, target2UP)\n",
    "    mergedDown = listMerger(target1DOWN, target2DOWN)\n",
    "\n",
    "\n",
    "    # change the ratios back to log2_fc\n",
    "    mergedUp = convert_ratio_to_log2fc(mergedUp)\n",
    "    mergedDown = convert_ratio_to_log2fc(mergedDown)\n",
    "\n",
    "    \n",
    "    # make sure we only have postive log2_fc tuples in the mergeUp list, and only negative log2_fc tuples in the mergeDown list, remove 0 \n",
    "    for gene in mergedUp:\n",
    "        if gene[1] < 0:\n",
    "            mergedDown.append(gene)\n",
    "            mergedUp.remove(gene)\n",
    "        elif gene[1] == 0:\n",
    "            mergedUp.remove(gene)\n",
    "    \n",
    "    for gene in mergedDown:\n",
    "        if gene[1] > 0:\n",
    "            mergedUp.append(gene)\n",
    "            mergedDown.remove(gene)\n",
    "        elif gene[1] == 0:\n",
    "            mergedDown.remove(gene)\n",
    "\n",
    "    return (mergedUp, mergedDown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for comparing a dual target to its individuals targets\n",
    "# compare the up and down regulated lists from the theoretical target merge to the real dual-target data (up and down regulated lists)\n",
    "# error checking so we only perform this on dual-targets that are defined\n",
    "\n",
    "def analyze(dualGeneP):\n",
    "\n",
    "    if dualGeneP not in dualGenes():\n",
    "        print(\"error: perturbation not found\")\n",
    "        return 1\n",
    "    \n",
    "    # get the 2 individual gene targets\n",
    "    target1 = ''\n",
    "    target2 = ''\n",
    "    for gene in genes():\n",
    "        if gene in dualGeneP:\n",
    "            if target1 == '':\n",
    "                target1 = gene\n",
    "            elif target2 == '':\n",
    "                target2 = gene\n",
    "   \n",
    "\n",
    "    theoreticalUp, theoreticalDown = merge(target1, target2)[0], merge(target1, target2)[1]\n",
    "    actualUp, actualDown = dualTarget(dualGeneP)[0], dualTarget(dualGeneP)[1]\n",
    "\n",
    "    # this returns 3 lists: \n",
    "    # genes that only get upregulated in the theoretical situation\n",
    "    # genes that only get upregulated in the real situation\n",
    "    # genes that get upregulated in both situations \n",
    "    theoU, realU, sameU = compare(theoreticalUp, actualUp)[0], compare(theoreticalUp, actualUp)[1], compare(theoreticalUp, actualUp)[2]\n",
    "\n",
    "    # this returns 3 lists: \n",
    "    # genes that only get downregulated in the theoretical situation\n",
    "    # genes that only get downregulated in the real situation\n",
    "    # genes that get downregulated in both situations \n",
    "    theoD, realD, sameD = compare(theoreticalDown, actualDown)[0], compare(theoreticalDown, actualDown)[1], compare(theoreticalDown, actualDown)[2]\n",
    "\n",
    "\n",
    "    #see if any theoretically go up, but actually go down\n",
    "    sameUD = compare(theoreticalUp, actualDown)[2]\n",
    "    #see if any theoretically go down, but actually go up\n",
    "    sameDU = compare(theoreticalDown, actualUp)[2]\n",
    "\n",
    "    return ((theoU), (realU), (sameU), (theoD), (realD), (sameD), (sameUD), (sameDU))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get number of theoretically DEGs \n",
    "\n",
    "def getTLength(dualPerturb):\n",
    "        \n",
    "    if dualPerturb not in dualGenes():\n",
    "        print(\"error: perturbation not found\")\n",
    "        return 1\n",
    "    \n",
    "    # get the 2 individual gene targets\n",
    "    target1 = ''\n",
    "    target2 = ''\n",
    "    for gene in genes():\n",
    "        if gene in dualPerturb:\n",
    "            if target1 == '':\n",
    "                target1 = gene\n",
    "            elif target2 == '':\n",
    "                target2 = gene\n",
    "    \n",
    "    upMergeL, downMergeL = len(merge(target1, target2)[0]), len(merge(target1, target2)[1])\n",
    "\n",
    "    return (upMergeL, downMergeL)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for getting single fold change\n",
    "def getSignleFC(dualPerturb, gene1):\n",
    "\n",
    "    if dualPerturb not in dualGenes():\n",
    "        print(\"error: perturbation not found\")\n",
    "        return 1\n",
    "    \n",
    "    # get the 2 individual gene targets\n",
    "    target1 = ''\n",
    "    target2 = ''\n",
    "    for gene in genes():\n",
    "        if gene in dualPerturb:\n",
    "            if target1 == '':\n",
    "                target1 = gene\n",
    "            elif target2 == '':\n",
    "                target2 = gene\n",
    "\n",
    "    value1 = 'NA'\n",
    "    value2 = 'NA'\n",
    "\n",
    "    # checking target1 up regulated genes\n",
    "    for element in geneTarget(target1)[0]:\n",
    "        if element[0] == gene1:\n",
    "            value1 = element[1]\n",
    "    \n",
    "    # checking target2 up regulated genes\n",
    "    for element in geneTarget(target2)[0]:\n",
    "        if element[0] == gene1:\n",
    "            value2 = element[1]\n",
    "\n",
    "    # checking target1 down regulated genes:\n",
    "    for element in geneTarget(target1)[1]:\n",
    "        if element[0] == gene1:\n",
    "            value1 = element[1]\n",
    "\n",
    "    # checking target2 down regulated genes:\n",
    "    for element in geneTarget(target2)[1]:\n",
    "        if element[0] == gene1:\n",
    "            value2 = element[1]\n",
    "    \n",
    "    return(value1, value2)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for testing\n",
    "\n",
    "def opposite(dualPerturb):\n",
    "    if dualPerturb not in dualGenes():\n",
    "        print(\"error: perturbation not found\")\n",
    "        return 1\n",
    "    \n",
    "    # get the 2 individual gene targets\n",
    "    target1 = ''\n",
    "    target2 = ''\n",
    "    for gene in genes():\n",
    "        if gene in dualPerturb:\n",
    "            if target1 == '':\n",
    "                target1 = gene\n",
    "            elif target2 == '':\n",
    "                target2 = gene\n",
    "\n",
    "    upMerge, downMerge = (merge(target1, target2)[0]), (merge(target1, target2)[1])\n",
    "    same = []\n",
    "    \n",
    "    for element in upMerge:\n",
    "        for element2 in downMerge:\n",
    "            if element[0] == element2[0]:\n",
    "                same.append((dualPerturb, element[0]))\n",
    "\n",
    "    return same\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "#to generate a text file comparing perturbations that target the same guide\n",
    "def perturbationsCompare():\n",
    "        \n",
    "    for gene in genes():\n",
    "        with open('comparingGuidesUP.txt', 'a') as file1:\n",
    "            for i in range(len(perturbations(gene))):\n",
    "                for j in range(i+1, len(perturbations(gene))):\n",
    "                    u1 = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'u')[0]\n",
    "                    u2 = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'u')[1]\n",
    "                    s = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'u')[2]\n",
    "                    print(\"comparing up regulation of\", perturbations(gene)[i],  \" and \", perturbations(gene)[j], \": \", file=file1)\n",
    "                    print(\"the unqiue\", len(u1), \" elements in \", perturbations(gene)[i], file=file1)\n",
    "                    print(u1, file=file1)\n",
    "                    print(\"the unique\", len(u2), \" elements in \", perturbations(gene)[j], file=file1)\n",
    "                    print(u2, file=file1)\n",
    "                    print(\"the shared\", len(s), \" elements: \", file=file1)\n",
    "                    print(s, \"\\n\\n\", file=file1)\n",
    "        \n",
    "        with open('comparingGuidesDOWN.txt', 'a') as file2:\n",
    "            for i in range(len(perturbations(gene))):\n",
    "                for j in range(i+1, len(perturbations(gene))):\n",
    "                    u1 = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'd')[0]\n",
    "                    u2 = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'd')[1]\n",
    "                    s = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'd')[2]\n",
    "                    print(\"comparing down regulation of\", perturbations(gene)[i],  \" and \", perturbations(gene)[j], \": \", file=file2)\n",
    "                    print(\"the unqiue\", len(u1), \" elements in \", perturbations(gene)[i], file=file2)\n",
    "                    print(u1, file=file2)\n",
    "                    print(\"the unique\", len(u2), \" elements in \", perturbations(gene)[j], file=file2)\n",
    "                    print(u2, file=file2)\n",
    "                    print(\"the shared\", len(s), \" elements: \", file=file2)\n",
    "                    print(s, \"\\n\\n\", file=file2)\n",
    "        \n",
    "        with open('comparingGuidesUP_DOWN.txt', 'a') as file3:\n",
    "            for i in range(len(perturbations(gene))):\n",
    "                for j in range(i+1, len(perturbations(gene))):\n",
    "                    u1 = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'd')[0]\n",
    "                    u2 = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'd')[1]\n",
    "                    s = compareM(perturbations(gene)[i], 'p', 'u', perturbations(gene)[j], 'p', 'd')[2]\n",
    "                    print(\"comparing up regulation of \", perturbations(gene)[i],  \" and down regulation of \", perturbations(gene)[j], \": \", file=file3)\n",
    "                    print(\"the unqiue\", len(u1), \" elements in \", perturbations(gene)[i], file=file3)\n",
    "                    print(u1, file=file3)\n",
    "                    print(\"the unique\", len(u2), \" elements in \", perturbations(gene)[j], file=file3)\n",
    "                    print(u2, file=file3)\n",
    "                    print(\"the shared\", len(s), \" elements: \", file=file3)\n",
    "                    print(s, \"\\n\\n\", file=file3)  \n",
    "        \n",
    "\n",
    "        with open('comparingGuidesDOWN_UP.txt', 'a') as file4:\n",
    "            for i in range(len(perturbations(gene))):\n",
    "                for j in range(i+1, len(perturbations(gene))):\n",
    "                    u1 = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'u')[0]\n",
    "                    u2 = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'u')[1]\n",
    "                    s = compareM(perturbations(gene)[i], 'p', 'd', perturbations(gene)[j], 'p', 'u')[2]\n",
    "                    print(\"comparing down regulation of \", perturbations(gene)[i],  \" and up regulation of \", perturbations(gene)[j], \": \", file=file4)\n",
    "                    print(\"the unqiue\", len(u1), \" elements in \", perturbations(gene)[i], file=file4)\n",
    "                    print(u1, file=file4)\n",
    "                    print(\"the unique\", len(u2), \" elements in \", perturbations(gene)[j], file=file4)\n",
    "                    print(u2, file=file4)\n",
    "                    print(\"the shared\", len(s), \" elements: \", file=file4)\n",
    "                    print(s, \"\\n\\n\", file=file4)\n",
    "\n",
    "\n",
    "perturbationsCompare()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "up regulated matches: \n",
      "  dual-target   gene  e log2_fc  a log2_fc     FC1     FC2  numEstim  numDual  \\\n",
      "0   ELF1-ELK1    CA1     0.2618     1.4162  0.2618     NaN        13       12   \n",
      "1   ELF1-E2F4   HBE1     0.2824     0.5193     NaN  0.2824        24        9   \n",
      "2   ELF1-E2F4   MT1X     0.6206     1.0896     NaN  0.6206        24        9   \n",
      "3    EGR1-YY1   DLK1     0.3366     0.4258     NaN  0.3366        65        6   \n",
      "4   GABPA-YY1   DLK1     0.3366     0.6108     NaN  0.3366        69       26   \n",
      "5    YY1-E2F4  KLHL8     0.2663     1.3177  0.2663     NaN        81       16   \n",
      "6    YY1-E2F4  MS4A3     0.2335     0.9063     NaN  0.2335        81       16   \n",
      "7    YY1-E2F4   DLK1     0.3366     0.5664  0.3366     NaN        81       16   \n",
      "8    YY1-E2F4   MT1X     0.6206     2.1446     NaN  0.6206        81       16   \n",
      "\n",
      "   numSame  \n",
      "0        1  \n",
      "1        2  \n",
      "2        2  \n",
      "3        1  \n",
      "4        1  \n",
      "5        4  \n",
      "6        4  \n",
      "7        4  \n",
      "8        4   \n",
      "\n",
      "\n",
      "down regulated matches: \n",
      "  dual-target       gene  e log2_fc  a log2_fc     FC1  FC2  numEstim  \\\n",
      "0  EGR1-NR2C2  HNRNPA1L2    -0.5272    -3.9198 -0.5272  NaN        44   \n",
      "\n",
      "   numDual  numSame  \n",
      "0       12        1   \n",
      "\n",
      "\n",
      "theoretically up regulated, actually down regulated: \n",
      "  dual-target   gene  e log2_fc  a log2_fc     FC1     FC2  numEstim  numDual  \\\n",
      "0  EGR1-GABPA  DHRS2     0.2956    -1.1995  0.2956     NaN        12       12   \n",
      "1    EGR1-YY1  DHRS2     0.2956    -1.5427  0.2956     NaN        65        9   \n",
      "2   EGR1-IRF1  DHRS2     0.2956    -1.5107  0.2956     NaN        20       18   \n",
      "3    YY1-IRF1  FOXL2     0.2599    -1.7048     NaN  0.2599        77       26   \n",
      "\n",
      "   numSame  \n",
      "0        1  \n",
      "1        1  \n",
      "2        1  \n",
      "3        1   \n",
      "\n",
      "\n",
      "theoretically down regulated, actually up regulated: \n",
      "  dual-target    gene  e log2_fc  a log2_fc     FC1     FC2  numEstim  \\\n",
      "0   CREB1-YY1  MS4A4A    -0.4928     2.7593     NaN -0.4928        44   \n",
      "1    YY1-IRF1  MS4A4A    -0.4928     2.7112 -0.4928     NaN        50   \n",
      "2    YY1-E2F4    GYPB    -0.2065     0.7910 -0.2065     NaN        60   \n",
      "3    YY1-E2F4  MS4A4A    -0.4928     3.0092 -0.4928     NaN        60   \n",
      "\n",
      "   numDual  numSame  \n",
      "0       11        1  \n",
      "1       28        1  \n",
      "2       16        2  \n",
      "3       16        2   \n",
      "\n",
      "\n",
      "checking opposite: \n",
      "('ELF1-E2F4', 'TFF1')\n",
      "('EGR1-YY1', 'RP11-433M22.2')\n",
      "('GABPA-YY1', 'ACRBP')\n",
      "('GABPA-IRF1', 'FRMD1')\n",
      "('YY1-E2F4', 'RP11-433M22.2')\n",
      "up regulated by YY1:  0.3366\n",
      "dual expression of DLK1 by NR2C2-YY1: 0\n"
     ]
    }
   ],
   "source": [
    "# main function \n",
    "\n",
    "def main():\n",
    "\n",
    "# base methods testing: \n",
    "\n",
    "    # this lists out all the target genes specified in the file, (the items in this list can be passed to the geneTarget function)\n",
    "    #print(genes())\n",
    "    \n",
    "\n",
    "    # this method lists out all the unique combinations of dual-gene targets in the specified file (path stored in var 'pathToDualGeneData')\n",
    "    #print(dualGenes())\n",
    "\n",
    "\n",
    "    # this method is used to find is a specific dual-gene perturbation exists in the data, specifying one of the genes in the pair\n",
    "    #print(findDual(\"ELF1\")) #example of a gene that is defined as a dual pair in the data\n",
    "    #print(findDual(\"ELF1Z\")) #example of a gene that is not defined as a dual pair in the data\n",
    "\n",
    "\n",
    "    # this method takes in a gene* name as a parameter and returns 2 lists which are the upregulated and downregulated genes as a result of the perturbation to the gene* \n",
    "    #print(\"upregulated genes: \")\n",
    "    #print(geneTarget(\"YY1\")[0])\n",
    "    #print(\"downregulated genes: \")\n",
    "    #print(geneTarget(\"YY1\")[1])\n",
    "\n",
    "\n",
    "    # this lists all perturbations that are associated with a specific gene \n",
    "    #print(len(perturbations(\"YY1\")), \" items found\")\n",
    "    #print(perturbations(\"YY1\"))\n",
    "\n",
    "\n",
    "    # this takes in a name of a specific perturbation and returns a list of upregulated genes, and a list of downregulated genes as a result of the perturbaiton \n",
    "    #print(\"upregulated genes: \")\n",
    "    #print(guideImpact(\"p-sgELF1-2\")[0])\n",
    "    #print(\"downregulated genes: \")\n",
    "    #print(guideImpact(\"p-sgELF1-2\")[1])\n",
    "\n",
    "\n",
    "    # this method does the same thing as geneTarget method and guideImpact methods, expect it takes in a specific dual-guide as a parameter\n",
    "    #print(dualTarget(\"ELF1-ELK1\"))\n",
    "\n",
    "\n",
    "\n",
    "#comparision methods testing:\n",
    "\n",
    "    # comparing specific situations (not for dual-targets)\n",
    "    # example: i want to compare which genes get upregulated when ELF1 is targeted, and which genes get upregulated when CREB1 is targeted\n",
    "    # mutation, type (either p or t) and direction (either u or d) as parameters \n",
    "    # note: the last list output (for shared elements) the elements in the list are tuples in the form (gene, log2_fc associated with first mutation parameter, log2_fc associated with second mutation parameter)\n",
    "    #compareM(\"CREB1\", 't', 'u', \"CREB1\", 't', 'u')\n",
    "    #compareM(\"CREB1\", 't', 'u', \"CREB1\", 't', 'u')\n",
    "    #compareM(\"YY1\", 't', 'u', \"CREB1\", 't', 'u')\n",
    "\n",
    "\n",
    "    # same thing as above put this example is for specific perturbations \n",
    "    # note YY1 only has two guides associated with it (found using print(perturbations(\"YY1\"))) good example is seeing what happens when you change the 3rd and 6th parameters\n",
    "    #compareM(\"p-sgYY1-3\", 'p', 'u', 'p-sgYY1-10', 'p', 'u') \n",
    "\n",
    "    # for all dual-target perturbations defined, a theoretical DGE list is generated if the two gene impacts added in a linear way\n",
    "    # so we obtain a list of theoretically up regulated genes (union of perturbation1 and perturbation2 upregulations), and same for the down regulated genes\n",
    "    # if there are any genes that are expressed in both lists, there log2_fc are combinded (log addition is performed)\n",
    "    # these lists are than compared to the up regulated and down regulated gene lists for the sitation in which the real dual-target perturbation was delivered\n",
    "    # up regulated matches means the gene listed was up regulated in both situtations (gene, theoretical expression, real expression)\n",
    "    \n",
    "    # this information is then stored in a dictonary, and then converted to a csv file \n",
    "    upMatchDict = {}\n",
    "    downMatchDict = {}\n",
    "    tUp_aDownDict = {}\n",
    "    tDown_aUpDict = {}\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    # for genes that were up regulated in both cases\n",
    "    for element in dualGenes():\n",
    "            if len(analyze(element)[2]) != 0: #ignoring the empty lines\n",
    "                upMatchDict[element] = analyze(element)[2]\n",
    "\n",
    "\n",
    "    with open('upMatchDict.csv', 'w', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        writer.writerow(['dual-target', 'gene', 'e log2_fc', 'a log2_fc', 'FC1', 'FC2', 'numEstim', 'numDual', 'numSame'])\n",
    "        for key, values in upMatchDict.items():\n",
    "            for value in values:\n",
    "                writer.writerow([key] + list(value) + [getSignleFC(key,value[0])[0]] + [getSignleFC(key,value[0])[1]] + [getTLength(key)[0]] + [len(dualTarget(key)[0])] + [len(values)])\n",
    "                \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # for genes that were down regulated in both vases\n",
    "    for element in dualGenes():\n",
    "        if len(analyze(element)[5]) != 0:\n",
    "            downMatchDict[element] = analyze(element)[5]\n",
    "\n",
    "    with open('downMatchDict.csv', 'w', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        writer.writerow(['dual-target', 'gene', 'e log2_fc', 'a log2_fc', 'FC1', 'FC2', 'numEstim', 'numDual', 'numSame'])\n",
    "        for key, values in downMatchDict.items():\n",
    "            for value in values:\n",
    "                writer.writerow([key] + list(value) + [getSignleFC(key,value[0])[0]] + [getSignleFC(key,value[0])[1]] + [getTLength(key)[1]] + [len(dualTarget(key)[1])]+ [len(values)])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # for genes that were up regulated theoretically, but actually down regulated with both guides present  \n",
    "    for element in dualGenes():\n",
    "        if len(analyze(element)[6]) != 0:\n",
    "            tUp_aDownDict[element] = analyze(element)[6]\n",
    "    \n",
    "    with open('tUp_aDownDict.csv', 'w', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        writer.writerow(['dual-target', 'gene', 'e log2_fc', 'a log2_fc', 'FC1', 'FC2', 'numEstim', 'numDual', 'numSame'])\n",
    "        for key, values in tUp_aDownDict.items():\n",
    "            for value in values:\n",
    "                writer.writerow([key] + list(value) + [getSignleFC(key,value[0])[0]] + [getSignleFC(key,value[0])[1]] + [getTLength(key)[0]] + [len(dualTarget(key)[1])] + [len(values)])\n",
    "\n",
    "\n",
    "\n",
    "    # for genes that were down regulated theoretically, but actually up regulated with both guides present\n",
    "    for element in dualGenes():\n",
    "        if len(analyze(element)[7]) != 0:\n",
    "            tDown_aUpDict[element] = analyze(element)[7]\n",
    "\n",
    "    with open('tDown_aUpDict.csv', 'w', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        writer.writerow(['dual-target', 'gene', 'e log2_fc', 'a log2_fc', 'FC1', 'FC2', 'numEstim', 'numDual', 'numSame'])\n",
    "        for key, values in tDown_aUpDict.items():\n",
    "            for value in values:\n",
    "                writer.writerow([key] + list(value) + [getSignleFC(key,value[0])[0]] + [getSignleFC(key,value[0])[1]] + [getTLength(key)[1]] + [len(dualTarget(key)[0])] + [len(values)])\n",
    "\n",
    "    # printing out whats in dictonaries\n",
    "    print(\"up regulated matches: \")\n",
    "    df1 = pd.read_csv('upMatchDict.csv')\n",
    "    print(df1, \"\\n\\n\")\n",
    "    print(\"down regulated matches: \")\n",
    "    df2 = pd.read_csv('downMatchDict.csv')\n",
    "    print(df2, \"\\n\\n\")\n",
    "    print(\"theoretically up regulated, actually down regulated: \")\n",
    "    df3 = pd.read_csv('tUp_aDownDict.csv')\n",
    "    print(df3, \"\\n\\n\")\n",
    "    print(\"theoretically down regulated, actually up regulated: \")\n",
    "    df4 = pd.read_csv('tDown_aUpDict.csv')\n",
    "    print(df4, \"\\n\\n\")\n",
    "\n",
    "\n",
    "    #any dual targets that have one of their targets up regulate a gene, but the other target down regulates it \n",
    "    print(\"checking opposite: \")\n",
    "    for element in dualGenes():\n",
    "        if len(opposite(element)) != 0:\n",
    "            print(opposite(element)[0])\n",
    "    \n",
    "    for element in geneTarget(\"NR2C2\")[0]:\n",
    "        if element[0] == \"DLK1\":\n",
    "            print(\"up regulated by NR2C2: \", element[1])\n",
    "    for element in geneTarget(\"NR2C2\")[1]:\n",
    "        if element[0] == \"DLK1\":\n",
    "            print(\"down regulated by NR2C2: \", element[1])\n",
    "    for element in geneTarget(\"YY1\")[0]:\n",
    "        if element[0] == \"DLK1\":\n",
    "            print(\"up regulated by YY1: \", element[1])\n",
    "    for element in geneTarget(\"YY1\")[1]:\n",
    "        if element[0] == \"DLK1\":\n",
    "            print(\"down regulated by YY1: \", element[1])\n",
    "    \n",
    "    \n",
    "    print(\"dual expression of DLK1 by NR2C2-YY1: \", end='')\n",
    "    found = False\n",
    "    for element in dualTarget(\"NR2C2-YY1\")[0]:\n",
    "        if element[0] == \"DLK1\":\n",
    "            print(element[1])\n",
    "            found = True\n",
    "    \n",
    "    for element in dualTarget(\"NR2C2-YY1\")[1]:\n",
    "            if element[0] == \"DLK1\":\n",
    "                print(element[1])\n",
    "                found = True\n",
    "    \n",
    "    if found == False:\n",
    "        print(\"0\")\n",
    "\n",
    "    return \n",
    "    \n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataRetrival",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
